src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1374,21 +1374,26 @@
         WHERE f.value IS NULL AND fv.value IS NULL")
 
   (log/info (trs "[5/8] Cleaning up duplicate null values..."))
-  (let [json-null-value-id (some-> (jdbc/query-to-vec "select id from fact_values
-                                                       where value_type_id = 5 and value = 'null'")
-                                   first
-                                   :id)
-        real-null-value-id (some-> (jdbc/query-to-vec "select id from fact_values
-                                                       where value_type_id = 4 and value is null")
-                                   first
-                                   :id)]
-    (if (and json-null-value-id real-null-value-id)
-      (do
-        (jdbc/do-prepared "update facts_transform set fact_value_id = (?) where fact_value_id = (?)"
-                          [real-null-value-id json-null-value-id])
-        (jdbc/do-prepared "delete from fact_values where id = (?)"
-                          [json-null-value-id]))
-      (log/info (trs "  ... none found"))))
+  (let [real-null-value-id (let [existing-id (some-> (jdbc/query-to-vec
+                                                      "select id from fact_values
+                                                       where value_type_id=4 and value is null")
+                                                     first
+                                                     :id)]
+                             (or existing-id
+                                 (-> (jdbc/query-to-vec
+                                      "insert into fact_values (value_type_id, value) values (4, null) returning id")
+                                     first
+                                     :id)))
+        bad-null-value-ids (->> (jdbc/query-to-vec "select id from fact_values
+                                                    where (value_type_id = 5 and value = 'null')
+                                                       or (value_type_id = 5 and value is null)
+                                                       or (value_type_id = 4 and value = 'null')")
+                                (map :id))]
+    (doseq [id bad-null-value-ids]
```

Comment:
Haven't though this through carefully, but wondered if we might be able to use an `update from` or subselect update to let pg do the work, and avoid having to worry about the size of the change set.

Reply:
This is post-deduplication, so there will be 3 rows max.